boot/cypress/cy_flash_pal/cy_flash_psoc6.c

```diff
@@ -0,0 +1,277 @@
+	uint32_t remStart, remEnd;
+	uint32_t rowIdxStart, rowIdxEnd, rowNum;
+	uint8_t  buff[CY_FLASH_SIZEOF_ROW];
+
+	addrStart = addr;
+	addrEnd   = addrStart + size;
+
+	/* find if area bounds are aligned to rows */
+	remStart = addrStart%CY_FLASH_SIZEOF_ROW;
+	remEnd   = addrEnd%CY_FLASH_SIZEOF_ROW;
+
+	/* find which row numbers are affected for full Erase */
+	rowIdxStart = addrStart/CY_FLASH_SIZEOF_ROW;
+	rowIdxEnd   = addrEnd/CY_FLASH_SIZEOF_ROW;
+
+	if(remStart != 0)
+	{/* first row is fragmented, move to next */
+		rowIdxStart++;
+	}
+
+	/* total number of rows for full erase */
+	rowNum = rowIdxEnd - rowIdxStart;
+	address = rowIdxStart*CY_FLASH_SIZEOF_ROW;
+
+	while(rowNum>0)
+	{
+		rc = Cy_Flash_EraseRow(address);
+		assert(rc == 0);
+		address += CY_FLASH_SIZEOF_ROW;
+		rowNum--;
+	}
+
+	/* if Start of erase area is unaligned */
+	if(remStart != 0)
+	{
+		/* first row is fragmented, shift left by one*/
+		rowIdxStart--;
+
+		/* find start address of fragmented row */
+		address = rowIdxStart*CY_FLASH_SIZEOF_ROW;
+
+		/* store fragmented row contents first */
+		memcpy((void *)buff, (const void*)address, remStart);
+
+		/* erase fragmented row */
+		rc = Cy_Flash_EraseRow(address);
+		assert(rc == 0);
+
+		/* write stored back */
+		rc = psoc6_flash_write_hal(buff, address, remStart);
+		assert(rc == 0);
+	}
+	/* if End of erase area is unaligned */
+	if(remEnd != 0)
+	{
+		/* find start address of fragmented row */
+		address = rowIdxEnd*CY_FLASH_SIZEOF_ROW;
+
+		/* store fragmented row contents first */
+		memcpy((void *)buff, (const void*)addrEnd, CY_FLASH_SIZEOF_ROW-remEnd);
+
+		/* erase fragmented row */
+		rc = Cy_Flash_EraseRow(address);
+		assert(rc == 0);
+
+		/* write stored back */
+		rc = psoc6_flash_write_hal(buff, addrEnd, CY_FLASH_SIZEOF_ROW-remEnd);
+		assert(rc == 0);
+	}
+	return rc;
+}
+
+/*******************************************************************************
+* Function Name: psoc6_flash_write_hal
+****************************************************************************//**
+*
+*  This function writes the data to the PSOC6's Flash. It will check the
+*  appropriate alignment of a start address and also perform an address range
+*  check based on the length before performing the write operation.
+*  This function performs memory compare and writes only row where there are new
+*  data to write.
+*
+*  \param addr:   Pointer to the buffer containing the data to be stored.
+*  \param data:   Pointer to the array or variable in the flash.
+*  \param len:    The length of the data in bytes.
+*
+* \return
+*  PSOC6_WR_SUCCESS                     A successful write
+*  PSOC6_WR_ERROR_INVALID_PARAMETER     At least one of the input parameters is invalid
+*  PSOC6_WR__ERROR_FLASH_WRITE          Error in flash Write
+*
+*******************************************************************************/
+int psoc6_flash_write_hal(uint8_t data[],
+							 uint32_t address,
+							 uint32_t len)
+{
```

Comment:
This also doesn't need to be this complex. MCUboot should never try writing to an area of flash that hasn't been erased. If you find that happening, it should cause a failure which can be investigated.

Reply:
We have opened a jira to rework this API and plan to submit this via another PR